add gamescope microcompositor support

aderumier opened the pull request:
This is a continuation of this work:

https://github.com/batocera-linux/batocera.linux/pull/13690

Gamescope is a microcompositor allowing to any emulator or process is a wayland surfarce,
with upscaling|downscaling support   with fsr or other modern techniques.

It's also add support for HDR for games supporting it. (mostly modern windows games).

For wine, it's really usefull, as currently, wine is using xrandr to switch the monitor resolution to the native game resolution with an automatic frequency (even if videomode is at batocer size defined, it'll can change it), and sometime it choose a bad unsupported frequency or resolution, and user can have a black screen or game not rendered full screen. (some 4k tvs have shit EDID).

This is also allowing some only 1080p games to run on steamdeck in 720p.  (Most of the modern arcade games have fixed internal resolution, and are not able to downscale)
or allow to run 16:9 in 4:3 arcade cabinet monitor

Gamescope is compatible with AMD (Mesa 20.3+), Intel (Mesa 21.2+) , Nvidia proprietary driver (515.43.04+). So no legacy driver for nvidia, as it need gbm support.

I have tested myself and some friend with : nvidia rtx2070, nvidia quadro, amd  rx 6600, Nvidia Quadro P1000, amd ryzen7 apu, steamdeck

aderumier commented:
> @aderumier good job, please mark it as draft until finished
> 
> a few of points:
> 
> 1. wayland should be optional as gamescope should run on xorg without wayland (afaik)
> 2. therefore make the dependencies based on the support for the board they're built against
> 
> * we don't want unnecessary packages enabled for an ecosystem which will affect other package builds
> 
> 3. please aim for .mk files to ideally stay with in the hash length for the package
> 4. tidy the config.in as well as you don't need to select all packages if they're optional
> 5. squash your commits :-)

I have only tested on xorg  with xwayland,  at minimum it need libwayland-client && libwaylab-server

gamescope is using wlroots internally and run only on pure wayland or xwayland(wit xorg)

I'm pretty noob with buildroot, so sorry, I'll try to reduce dependencies at minimum again. (I had already tried to reduce it)

(the depend in .mk are ok , but i'm not sure in Config.in )

BR2_PACKAGE_WAYLAND_PROTOCOLS  && BR2_PACKAGE_XWAYLAND  are needed at minimum, I'm not sure about BR2_PACKAGE_WAYLAND


I'll do test again.

aderumier commented:
to be more precise, gamescope is opening a new wayland session over current Xorg with only the game running in this session fullscreen.   And the game inside gamescope, can run with Xwayland  or native wayland if it support it).

dmanlfc commented:
@aderumier i think you're confused. xwayland is for backward compatibility with wayland compositors. are you suggesting xorg -> wayland -> _> gamescope -> xwayland?
if wayland is truly needed then this opens up a can or worms for the xorg only x86_64 build.
are you sure you just don't need wlroots with possibly wayland protocols?

i will look at their source code...

dmanlfc commented:
ok we have a problem, it does look like wayland is needed.
this will break the build and set the wrong expectations for older nvidia card users.
we can accept it for the newer zen3 images but not the x86_64 one.

dmanlfc commented:
ok built it, yes it needs xwayland as a dependency.

like this @aderumier 
```
################################################################################
#
# gamescope
#
################################################################################

GAMESCOPE_VERSION = 3.16.17
GAMESCOPE_SITE = https://github.com/ValveSoftware/gamescope
GAMESCOPE_SITE_METHOD = git
GAMESCOPE_GIT_SUBMODULES=YES

GAMESCOPE_DEPENDENCIES += hwdata libdisplay-info libdecor luajit seatd
GAMESCOPE_DEPENDENCIES += vulkan-headers vulkan-loader wayland wayland-protocols 
GAMESCOPE_DEPENDENCIES += xlib_libX11 xlib_libXmu xlib_libXres xwayland

GAMESCOPE_CONF_OPTS = --wrap-mode=default
GAMESCOPE_CONF_OPTS += -Dbenchmark=disabled
GAMESCOPE_CONF_OPTS += -Denable_openvr_support=false
GAMESCOPE_CONF_OPTS += -Dinput_emulation=disabled

ifeq ($(BR2_PACKAGE_LIBAVIF),y)
GAMESCOPE_DEPENDENCIES += libavif
GAMESCOPE_CONF_OPTS += -Davif_screenshots=enabled
else
GAMESCOPE_CONF_OPTS += -Davif_screenshots=disabled
endif

ifeq ($(BR2_PACKAGE_LIBDRM),y)
GAMESCOPE_DEPENDENCIES += libdrm
GAMESCOPE_CONF_OPTS += -Ddrm_backend=enabled
else
GAMESCOPE_CONF_OPTS += -Ddrm_backend=disabled
endif

ifeq ($(BR2_PACKAGE_PIPEWIRE),y)
GAMESCOPE_DEPENDENCIES += pipewire
GAMESCOPE_CONF_OPTS += -Dpipewire=enabled
else
GAMESCOPE_CONF_OPTS += -Dpipewire=disabled
endif

ifeq ($(BR2_PACKAGE_SDL2),y)
GAMESCOPE_DEPENDENCIES += sdl2
GAMESCOPE_CONF_OPTS += -Dsdl2_backend=enabled
else
GAMESCOPE_CONF_OPTS += -Dsdl2_backend=disabled
endif

define GAMESCOPE_INSTALL_TARGET_CMDS
	mkdir -p $(TARGET_DIR)/usr/bin
	$(INSTALL) -D $(@D)/build/src/gamescope $(TARGET_DIR)/usr/bin/gamescope
	$(INSTALL) -D $(@D)/build/src/gamescopereaper $(TARGET_DIR)/usr/bin/gamescopereaper
	$(INSTALL) -D $(@D)/build/src/gamescopestream $(TARGET_DIR)/usr/bin/gamescopestream
	$(INSTALL) -D $(@D)/build/src/gamescopectl $(TARGET_DIR)/usr/bin/gamescopectl
endef

$(eval $(meson-package))
```
and...

```
config BR2_PACKAGE_GAMESCOPE
    bool "gamescope"
    select BR2_PACKAGE_HWDATA
    select BR2_PACKAGE_LIBDECOR
    select BR2_PACKAGE_LIBDISPLAY_INFO
    select BR2_PACKAGE_LIBX11
    select BR2_PACKAGE_LIBXCB
    select BR2_PACKAGE_LIBXRES
    select BR2_PACKAGE_LUAJIT    if !BR2_PACKAGE_LUA
    select BR2_PACKAGE_SEATD
    select BR2_PACKAGE_SEATD_DAEMON
    select BR2_PACKAGE_VULKAN_HEADERS
    select BR2_PACKAGE_VULKAN_LOADER
    select BR2_PACKAGE_WAYLAND
    select BR2_PACKAGE_WAYLAND_PROTOCOLS
    select BR2_PACKAGE_WLROOTS
    select BR2_PACKAGE_WLROOTS_X11
    select BR2_PACKAGE_XLIB_LIBXMU
    select BR2_PACKAGE_XWAYLAND
    help
      SteamOS session compositing window manager

      https://github.com/Plagman/gamescope

```

dmanlfc commented:
easier said than done unfortunately.
until such time i still think `xorg -> gamescope -> wayland -> xwayland` (and it's probably more complicated than that) is plain stupid. not to mention messy.
running `wayland -> gamescope` and occasionally when you need it for compatibility  `wayland -> gamescope -> xwayland` is much nicer.
hence bringing it into zen3 which is wayland by design is a much better starting point.

aderumier commented:
> easier said than done unfortunately. until such time i still think `xorg -> gamescope -> wayland -> xwayland` (and it's probably more complicated than that) is plain stupid. not to mention messy. running `wayland -> gamescope` and occasionally when you need it for compatibility `wayland -> gamescope -> xwayland` is much nicer. hence bringing it into zen3 which is wayland by design is a much better starting point.

no, it's more

```
-->xorg session--->gamescope  -> xwayland---> X11 game
                               -> wayland  native game
```


They are not "wayland" server,  wayland is a lib, and the "server" is the compositor : gamescope, sway, ...

without xorg, you could used gamescope nested behind sway, or directly gamescope in standalone, managing the whole sessions


nested
```
-->sway--->gamescope nested -> xwayland---> X11 game
                            -> wayland  native game
```

standalone
```
-->gamescope session -> xwayland---> X11 game
                      -> wayland  native game
```


(I think that steamdeck steamos use standalone model)

aderumier commented:
```
BR2_PACKAGE_BATOCERA_WAYLAND && BR2_PACKAGE_BATOCERA_VULKAN` to possibly test higher-end arm boards too but that's up to Valve... SteamDeck 2? :-)
```
technically, they are already doing it in their new vr headset running on snapdragon.
But running windows x86  games through an arm emulator... ^_^


Note that they are plenty of usecase for gamescope outside handled devices

- a lot of arcade cabinets with 4:3 screen can't run modern (windows based) arcade games,

- some 4k tv with bad edid are doing shit with wine . wine is using xrandr api to switch monitor resolution && frequency to game native resolution, even if you have fixed it in ES videomode and for old games with 640x480 , 800x600 this is breaking display (blackscreen, not fully displayed games,...) of theses monitor.

Redemp commented:
> @Redemp all the OS's you mentioned run Wayland by default and are forks of or provide an OS similar to SteamOS from which gamescope was designed for. I don't know how many times you wish me to say this but bringing in a bunch of dependencies for wayland will have a knock-on effect for other packages in the x86_64 build ecosystem. effectively building support for wayland into said packages also. this may have a negative impact and I will be the one who would need to tidy it up. I have no doubt this works in your testing but I have to think of the bigger picture and also support. Could we raise a support case for gamescope on xorg? these are the things i need to consider. since your driver identification is not added to this PR all I can say is that this would be an ideal approach before committing to x86_64. until such time we can bring this PR into zen3 which is more controlled & then expand from there.
> 
> the PR is still incomplete and should be marked as draft until things are tidied up.
> 
> but for now `select BR2_PACKAGE_GAMESCOPE if BR2_PACKAGE_BATOCERA_TARGET_ZEN3`
> 
> in future we can do `select BR2_PACKAGE_GAMESCOPE if BR2_PACKAGE_BATOCERA_XWAYLAND && BR2_PACKAGE_BATOCERA_WAYLAND && BR2_PACKAGE_BATOCERA_VULKAN` to possibly test higher-end arm boards too but that's up to Valve... SteamDeck 2? :-)



I did read what you wrote, and I completely understand the worry about older NVIDIA cards and not wanting to mislead those users.

My point is that this does not have to mean "no Gamescope at all" on the generic x86_64 image.

We already ship a lot of features that are not universally supported on every GPU or board:

- Vulkan backends inside emulators  
- DXVK for Wine  
- Demanding Vulcan shader presets  
- Various threaded / async video options  

On unsupported hardware if a user forces them anyway the renderer fails and they are simply dropped back to EmulationStation. Nothing is permanently broken, and users quickly learn "this option does not work on my card".

Gamescope on x86_64 behaves in the same way in practice:

- On my R9 380X (amdgpu, Mesa) it runs fine as a nested compositor on Xorg.  
- If you try it on a GPU or driver that does not meet the requirements, Gamescope fails and ES comes back.  

The problem you are describing ("wrong expectations for older NVIDIA") is a runtime capability problem, not a reason to exclude Gamescope from the generic image entirely.


To be clear, this is also why I do not think it makes sense to restrict Gamescope only to a future Wayland based, AMD handled build. That would be an unnecessarily narrow target for a feature that already works today on the generic x86_64 image:

- It would exclude a lot of perfectly capable AMD and NVIDIA setups that can already run Gamescope on Xorg without issues.  


I hope this does not come across the wrong way, but from my perspective it currently feels a bit like the "Little Red Hen" story: a lot of work has already gone into getting Gamescope working reliably on the generic x86_64 build, exactly so existing users on modern displays and CRT setups can benefit from it. Restricting it only to a future Wayland based AMD image would be like baking the bread, but then telling most of the people who helped grow the wheat that they cannot have a slice.

My only real ask is that we do not throw away a working, tested path for the generic image when a capability based approach can keep expectations sane for unsupported GPUs and still let all the hardware that already can run Gamescope actually use it.

digitech2308 commented:
Man im so excited about this, want to thank all the ones that are involved for the hard work, valious time and dedication. Me and my friends sometime play games that only works at fixed resolutions and this would be god send on all kinds of displays, i have a hd crt and this would be a option that i dont have anywhere else, heck im going to send a message to RetroRGB and MarcoRetro they both have YT channels related to gaming and im sure they would be very interested in the progress of gamescope.

ZFEbHVUE commented:
I agree with the previous comments, because having an option to control resolutions is a significant advantage for Wine and other emulators.

bryanforbes commented on package/batocera/core/batocera-configgen/configgen/configgen/utils/gamescope.py:
```diff
@@ -0,0 +1,103 @@
+
+_logger = logging.getLogger(__name__)
+
+@contextmanager
```
Why are you making this a context manager? You're not using it as one and you never `yield` from it.

bryanforbes commented on package/batocera/core/batocera-configgen/configgen/configgen/utils/gamescope.py:
```diff
@@ -0,0 +1,103 @@
+    if not system.config.get_bool("gamescope"):
+        return
+
+    gamescope_cmd = ["/usr/bin/gamescope"]
```
This should be named `gamescope_arguments` since it's not just the command

bryanforbes commented on package/batocera/core/batocera-configgen/configgen/configgen/utils/gamescope.py:
```diff
@@ -0,0 +1,103 @@
+        ])
+
+    # Retrieve nested refresh rate.
+    # Default 60 if undefined
```
Should we provide the default? If gamescope decides to change their default, our default will change unless we use one here.

kl4ng3r commented:
Want to add my +1 support for this initiative. Batocera would make a great OS for handheld systems. A lot o them use portrait native displays ... which means their image is rotated by 90 degrees by default and in driver you have to rotate the screen. This breaks compatibility with a lot of Wine games, both old and new. Also having the ability to lock in a particular refresh rate like 40 fps / hz would be another godsend for users of x86 handhelds, it would help save on battery life where a lot of these handhelds have only 1 supported refresh rate that you can vsync games to and thats 60 or 120 hz. This is overkill for a handheld. This feature would allow for the customizability to have the screen oriented at a desired resolution and frequency.

manfredi90 commented:
I'd like to add my +1 support for this initiative too. I'm especially interested in the potential implications using batocera with CRTs, but overall it would be a really significant development...

tenyuhuang commented:
Definitely would want this for x86_64 devices in general instead of a confined set of hardware. Low-res LCDs and CRTs, which are ideal for low res content display, would benefit a lot from this without having to abide to hardware confinements (i.e. you should be able to use whatever modern hardware you want to).

aderumier commented:
I have updated the PR.

- merged commits
- change Config.ini && .mk file  like dmanlfc proposal
- fix Bryan comments
- added a detection of gamescope support through driver gbm EGL_KHR_platform_gbm support

Redemp commented:
> I have updated the PR.
> 
> * merged commits
> * change Config.ini && .mk file  like dmanlfc proposal
> * fix Bryan comments
> * added a detection of gamescope support through driver gbm EGL_KHR_platform_gbm support

Thanks a lot for the update @aderumier this looks really good.

The new GBM / `EGL_KHR_platform_gbm` detection is exactly the kind of capability check we were talking about. 

So this should directly address one of the main concerns that was raised about older unsupported NVIDIA cards, while still keeping the door open for the hardware that already runs Gamescope without issues.

crcerror commented:
I've a `AMD Radeon HD 8280E` - it should support wayland but it is very vey old .... ? So I'm eager to test yet

skysix2 commented:
Want to add my +1 support for this initiative, Other emulators that would benefit from Gamescope are Mugen (obviously Wine) and Ikemen, there are often games that offer resolutions/aspect ratios that do not match well with modern monitors.

Tovarichtch commented:
More comments from unknown users won't help merge the PR. Commits will.

Please fix what's required to be done, thanks.

dmanlfc commented:
@aderumier you haven't selected gamescope in any package or batocera-system Config.in, you haven't added it to the `utils` in `Config.in`.
therefore this won't be added to any image in the current state.
i did ask you multiple times to mark it as draft until you finished, this hasn't happened either. so I have no idea whether you think it's ready or not. as it stands for me currently it's not ready.
because of these basics I therefore know you haven't built a x86_64 cleanbuild image from scratch to determine the effects of bringing in the dependencies of gamescope to the x86_64 image and it's possible effects on emulators.

so if you want this added quickly, fix the basics outlined above with the selection for the zen3 image currently and then we can look at x86_64 after.
if you want to add x86_64 support within this single PR then you need to handle legacy x86_64 drivers without gbm & vulkan support. i don't see this code yet (again mark it draft).

welcome to the wider ecosystem :-)

i'm also not sure on the `es_features.yml` changes but that's more of a design discussion the bato dev team need to have.

aderumier commented:
> @aderumier what's the status?

I didn't have time to work on yet (super busy at work). I'll try to test next week. (I need to prepare a big vm for building the whole image,as it's pretty long)

aderumier committed:
add gamescope support

when enabled, gamescope is prefixing the emulator commandline.

ex: gamescope -W 3840 -H 2160 -f -- /usr/bin/retroarch ....

default gamescope output resolution is set to current monitor resolution

external libs are statically linked to avoid conflict with system libs as
gamescope need specific libs version.

Xwayland binary is the only addition (with gamescope binary) to classic X86 xorg,
but don't conflict or impact any other packages.
